File: 01/a.cpp
#include <algorithm>
#include <array>
#include <iostream>
#include <iterator>
#include <vector>


using li = long int;
using ll = long long;


int two_sum(std::vector<ll> &arr, size_t left, int target){
    std::size_t right = arr.size() - 1;

    int cur;
    while (left < right)
    {
        cur = arr.at(left) + arr.at(right);
        if (cur < target)
            ++left;
        else if (cur > target)
            --right;
        else
            return arr.at(left) * arr.at(right);
    }
    return -1;
}


int three_sum(std::vector<ll> &arr, int target){
    using std::size_t;
    for (size_t idx{ 0 }; idx < arr.size(); ++idx)
    {
        int two_res{ two_sum(arr, idx + 1, target - arr[idx]) };
        if (two_res > 0)
            return arr[idx] * two_res;
    }
    return -1;
}


int main(){
    using std::cout;
	std::vector<ll> arr( std::istream_iterator<long int>(std::cin), {});
    std::sort(arr.begin(), arr.end());
    cout << "Res a: " << two_sum(arr, 0, 2020) << "\n";
    cout << "Res b: " << three_sum(arr, 2020) << "\n";
    return 0;
}

File: 01/a.py
#!/usr/bin/env python3

def get_nums():
    nums = []
    with open("in.txt") as text:
        for line in text:
            nums.append(int(line.strip('\n')))
    return nums

def two_sum():
    nums = get_nums()
    mem = set()
    for num in nums:
        a = 2020 - num
        if a in mem:
            print(num * a)
            return
        mem.add(num)

def two_sum_limits(sorted_nums, start, target):
    left, right = start, len(sorted_nums) - 1
    while left < right:
        s = sorted_nums[left] + sorted_nums[right]
        if s < target:
            left += 1
        elif s > target:
            right -= 1
        else:
            return sorted_nums[left] * sorted_nums[right]
    return None

def three_sum():
    nums = get_nums()
    nums.sort()
    for idx in range(len(nums)):
        two = two_sum_limits(nums, idx+1, 2020 - nums[idx])
        if two is not None:
            print(nums[idx] * two)


if __name__ == "__main__":
    two_sum()
    three_sum()
